src/hotspot/cpu/s390/templateTable_s390.cpp

```diff
@@ -2816,124 +2900,152 @@ void TemplateTable::jvmti_post_field_mod(Register cache,
   const Register oopStore_tmp3 = Z_R0_scratch;
+#ifdef ASSERT
+  const Register br_tab_temp   = Z_R0_scratch; // for branch table verification code only
+#endif
 
-  resolve_cache_and_index(byte_no, cache, index, sizeof(u2));
+/*
+ *  Register usage and life range
+ *
+ *  cache, index          : short-lived. Their life ends after load_resolved_field_entry.
+ *  obj (overwrites cache): very short-lived, Combined with off immediately.
+ *  off (overwrites index): long-lived, Used in branch table entries.
+ *  flags                 : long-lived, Has to survive until the end to determine volatility.
+ *  br_tab                : short-lived, Only used to address branch table, and for verification in BTB_BEGIN macro.
+ *  tos_state             : short-live, Only used to index the branch table entry.
+ *  bc_reg                : short-lived, Used as work register in patch_bytecode.
+*/
+  resolve_cache_and_index_for_field(byte_no, cache, index);
   jvmti_post_field_mod(cache, index, is_static);
-  load_field_cp_cache_entry(obj, cache, index, off, flags, is_static);
-  // begin of life for:
-  //   obj, off   long life range
-  //   flags      short life range, up to branch into branch table
-  // end of life for:
-  //   cache, index
+  load_resolved_field_entry(obj, cache, tos_state, off, flags, is_static);
+
+  // Displacement is 0. No need to care about limited displacement range.
+  const Address field(fieldAddr);
+  __ lgr_if_needed(fieldAddr, off);
+
+  /*
+   * In the static case, we can calculate the final field address easily.
+   * Do so to occupy only one non-volatile register
+   * ---------------------
+   * In the non-static case, we preset fieldAddr with the field offset.
+   * The object address is available only later. It is popped from stack.
+   * see pop_and_check_object(obj);
+   */
+  if (is_static) {
+    __ z_agr(fieldAddr, obj);
+  }
 
-  const Address field(obj, off);
-  Label is_Byte, is_Bool, is_Int, is_Short, is_Char,
+  Label is_Byte, is_Bool,  is_Int,    is_Short, is_Char,
         is_Long, is_Float, is_Object, is_Double;
-  Label is_badState8, is_badState9, is_badStateA, is_badStateB,
-        is_badStateC, is_badStateD, is_badStateE, is_badStateF,
-        is_badState;
+  Label is_badState,  is_badState9, is_badStateA, is_badStateB,
+        is_badStateC, is_badStateD, is_badStateE, is_badStateF;
   Label branchTable, atosHandler, Done;
   bool  do_rewrite   = !is_static && (rc == may_rewrite);
   bool  dont_rewrite = (is_static || (rc == may_not_rewrite));
 
   assert(do_rewrite == !dont_rewrite, "Oops, code is not fit for that");
-
-  assert(btos == 0, "change code, btos != 0");
+  assert((btos == 0) && (atos == 8), "change branch table! ByteCodes may have changed");
 
 #ifdef ASSERT
-  const unsigned int bsize = is_static ? BTB_MINSIZE*1 : BTB_MINSIZE*4;
+  const unsigned int bsize = is_static ? BTB_MINSIZE*1 : BTB_MINSIZE*8; // TODO: why code size is increasing ?
```

Comment:
TODO should get resolved.

Reply:
Multiple test failures appeared after the code change with error: ` guarantee(len <= bsize) failed: block too large `, reducing it back is not a option as of now, but probably in future it could be shrink back to normal. So I've planned to deal with separately, as build is broken for now.

Reply:
You don't need to make it smaller, now, but please improve the comment or file a JBS issue for that and remove the comment.

Reply:
Opened Issue for future: https://bugs.openjdk.org/browse/JDK-8317673